modes/parkour/moderation.lua

```diff
@@ -0,0 +1,447 @@
+local files = {
+	[1] = 20, -- maps, ranks, chats
+	[2] = 22 -- lowmaps, banned
+}
+
+local loaded = false
+local saving = {}
+local to_do = {}
+local ban_request = {}
+
+local file_actions = {
+	ban_change = {2, true, function(data, id, value, old)
+		id = tostring(id)
+		if not old or tonumber(old) == data.banned[id] then
+			data.banned[id] = value
+		end
+	end},
+
+    high_map_change = {1, true, function(data, map, add)
+		for index = #data.maps, 1, -1 do
+			if data.maps[index] == map then
+				if not add then
+					table.remove(data.maps, index)
+				end
+				return
+			end
+		end
+
+		if add then
+			data.maps[#data.maps + 1] = map
+		end
+	end},
+
+	low_map_change = {2, true, function(data, map, add)
+		for index = 1, #data.lowmaps do
+			if data.lowmaps[index] == map then
+				if not add then
+					table.remove(data.lowmaps, index)
+				end
+				return
+			end
+		end
+
+		if add then
+			data.lowmaps[#data.lowmaps + 1] = map
+		end
+	end},
+
+    modify_rank = {1, true, function(data, player, newrank)
+		data.ranks[player] = newrank
+	end},
+}
+
+local function schedule(action, arg1, arg2, arg3)
+	to_do[#to_do + 1] = {file_actions[action], arg1, arg2, arg3}
+	next_file_check = os.time() + 4000
+end
+
+
+onEvent("FileLoaded", function(file, data)
+	data = filemanagers[file]:load(data)
+	file = tonumber(file)
+	local save = false
+
+	local action
+	for index = 1, #to_do do
+		action = to_do[index][1]
+		if files[action[1]] == file then
+			if action[2] then
+				save = true
+			end
+
+			action[3](data, to_do[index][2], to_do[index][3])
+			to_do[index] = nil
+		end
+	end
+	for index = #to_do, 1, -1 do
+		if not to_do[index] then
+			table.remove(to_do, index)
+		end
+	end
+
+	eventGameDataLoaded(data)
+
+	if data.banned then
+		local now = os.time()
+		local count, to_remove = 0
+
+		-- wipe expired bans
+		for id, value in next, data.banned do
+			if value > 1 and now >= value then
+				if count == 0 then
+					count = 1
+					to_remove = {id}
+				else
+					count = count + 1
+					to_remove[count] = id
+				end
+			end
+		end
+
+		for index = 1, count do
+			data.banned[to_remove[index]] = nil
+		end
+	end
+
+	if save then
+		eventSavingFile(tostring(file), data)
```

Comment:
And you're also saving the file twice. I'd recommend editing `files.lua` at this point.